src/utils/translator.rs

```diff
@@ -0,0 +1,494 @@
+	Function,
+	Table,
+	Memory,
+	Tag,
+	Global,
+	Type,
+	Data,
+	Element,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Hash, Eq, PartialEq, Copy, Clone)]
+pub enum ConstExprKind {
+	Global,
+	ElementOffset,
+	ElementFunction,
+	DataOffset,
+	TableInit,
+}
+
+pub trait Translator {
+	fn as_obj(&mut self) -> &mut dyn Translator;
+
+	fn translate_type_def(&mut self, ty: Type, s: &mut TypeSection) -> Result<()> {
+		type_def(self.as_obj(), ty, s)
+	}
+
+	fn translate_import(&mut self, import: Import, s: &mut ImportSection) -> Result<()> {
+		import_def(self.as_obj(), import, s)
+	}
+
+	fn translate_table_type(
+		&mut self,
+		ty: &wasmparser::TableType,
+	) -> Result<wasm_encoder::TableType> {
+		table_type(self.as_obj(), ty)
+	}
+
+	fn translate_memory_type(
+		&mut self,
+		ty: &wasmparser::MemoryType,
+	) -> Result<wasm_encoder::MemoryType> {
+		memory_type(self.as_obj(), ty)
+	}
+
+	fn translate_global_type(
+		&mut self,
+		ty: &wasmparser::GlobalType,
+	) -> Result<wasm_encoder::GlobalType> {
+		global_type(self.as_obj(), ty)
+	}
+
+	fn translate_tag_type(&mut self, ty: &wasmparser::TagType) -> Result<wasm_encoder::TagType> {
+		tag_type(self.as_obj(), ty)
+	}
+
+	fn translate_ty(&mut self, t: &wasmparser::ValType) -> Result<ValType> {
+		ty(self.as_obj(), t)
+	}
+
+	fn translate_refty(&mut self, t: &wasmparser::RefType) -> Result<RefType> {
+		refty(self.as_obj(), t)
+	}
+
+	fn translate_heapty(&mut self, t: &wasmparser::HeapType) -> Result<HeapType> {
+		heapty(self.as_obj(), t)
+	}
+
+	fn translate_global(&mut self, g: Global, s: &mut GlobalSection) -> Result<()> {
+		global(self.as_obj(), g, s)
+	}
+
+	fn translate_export_kind(
+		&mut self,
+		g: wasmparser::ExternalKind,
+	) -> Result<wasm_encoder::ExportKind> {
+		export_kind(self.as_obj(), g)
+	}
+
+	fn translate_export(
+		&mut self,
+		e: &wasmparser::Export,
+		sec: &mut wasm_encoder::ExportSection,
+	) -> Result<()> {
+		export(self.as_obj(), e, sec)
+	}
+
+	fn translate_const_expr(
+		&mut self,
+		e: &wasmparser::ConstExpr<'_>,
+		_ty: &wasmparser::ValType,
+		ctx: ConstExprKind,
+	) -> Result<wasm_encoder::ConstExpr> {
+		const_expr(self.as_obj(), e, ctx)
+	}
+
+	fn translate_element(
+		&mut self,
+		e: wasmparser::Element<'_>,
+		s: &mut ElementSection,
+	) -> Result<()> {
+		element(self.as_obj(), e, s)
+	}
+
+	fn translate_data(&mut self, d: wasmparser::Data<'_>, s: &mut DataSection) -> Result<()> {
+		data(self.as_obj(), d, s)
+	}
+
+	fn translate_code(&mut self, body: FunctionBody<'_>, s: &mut CodeSection) -> Result<()> {
+		code(self.as_obj(), body, s)
+	}
+
+	fn translate_op(&mut self, e: &Operator<'_>) -> Result<Instruction<'static>> {
+		op(self.as_obj(), e)
+	}
+
+	fn translate_block_type(&mut self, ty: &wasmparser::BlockType) -> Result<BlockType> {
+		block_type(self.as_obj(), ty)
+	}
+
+	fn translate_memarg(&mut self, arg: &wasmparser::MemArg) -> Result<MemArg> {
+		memarg(self.as_obj(), arg)
+	}
+
+	fn remap(&mut self, item: Item, idx: u32) -> Result<u32> {
```

Comment:
What is remap?

Reply:
Not sure about the author's intention. Maybe he wanted to express that conversion between wasmparser and wasm_encoder is happening. 
Maybe he predicts that more logic will be needed here in the future.

Reply:
Will remove it